sources/packages/backend/apps/api/src/auth/guards/maintenance-mode.guard.ts

```diff
@@ -0,0 +1,74 @@
+import {
+  Injectable,
+  CanActivate,
+  ExecutionContext,
+  ServiceUnavailableException,
+} from "@nestjs/common";
+import { Reflector } from "@nestjs/core";
+import { AuthorizedParties, IUserToken } from "..";
+import { ConfigService } from "@sims/utilities/config";
+import { ALLOW_DURING_MAINTENANCE_MODE_KEY } from "../decorators/allow-during-maintenance-mode.decorator";
+
+/**
+ * Guard that prevents access to the API when the system or a specific client portal is in maintenance mode.
+ * Maintenance mode is used during planned system upgrades, emergency fixes, or other periods when access must be restricted to ensure data integrity and user safety.
+ * The guard checks both global maintenance mode and client-type-specific maintenance flags, and when triggered, throws a ServiceUnavailableException to inform users that the system is temporarily unavailable.
+ */
+@Injectable()
+export class MaintenanceModeGuard implements CanActivate {
+  constructor(
+    private readonly configService: ConfigService,
+    private readonly reflector: Reflector,
+  ) {}
+
+  canActivate(context: ExecutionContext): boolean {
+    const allowDuringMaintenanceMode =
+      this.reflector.getAllAndOverride<boolean>(
+        ALLOW_DURING_MAINTENANCE_MODE_KEY,
+        [context.getHandler(), context.getClass()],
+      );
+
+    // If the route has the allowed during maintenance mode decorator, bypass the guard.
+    if (allowDuringMaintenanceMode) {
+      return true;
+    }
+
+    // Get authorized party from the user token to determine maintenance mode status.
+    const { user } = context.switchToHttp().getRequest();
+    const userToken = user as IUserToken;
+    if (this.isMaintenanceModeEnabled(userToken?.azp)) {
+      throw new ServiceUnavailableException(
+        "Service temporarily unavailable for maintenance. Please try again later or contact support if the issue persists.",
+      );
+    }
+
+    return true;
+  }
+
+  /**
+   * Determines if maintenance mode is enabled for the specified authorized party.
+   * Checks global first, then client-type-specific flags.
+   * @param authorizedParty authorized party type to be checked.
+   * @returns True if maintenance mode is enabled, otherwise false.
+   */
+  private isMaintenanceModeEnabled(
+    authorizedParty?: AuthorizedParties,
+  ): boolean {
+    // Check global maintenance mode first.
+    if (this.configService.maintenanceMode) return true;
+    // If no authorized party is provided, maintenance mode cannot be determined.
+    if (!authorizedParty) return false;
```

Comment:
Is this `if` required? Would it not be taken care of at line 72, or is it just to return early and did not check anything further?

Reply:
not required but the early return makes it a bit clearer